gate-core/src/main/java/com/netflix/spinnaker/gate/services/PermissionService.java

```diff
@@ -133,22 +136,31 @@ public Set<Role.View> getRoles(String userId) {
       var permission = permissionEvaluator.getPermission(userId);
       var roles = permission != null ? permission.getRoles() : null;
       return roles != null ? roles : Set.of();
-    } catch (RetrofitError e) {
+    } catch (SpinnakerServerException e) {
       throw UpstreamBadRequest.classifyError(e);
     }
   }
 
   List<UserPermission.View> lookupServiceAccounts(String userId) {
     try {
       return extendedFiatService.getUserServiceAccounts(userId);
-    } catch (RetrofitError re) {
-      var response = re.getResponse();
-      if (response != null && response.getStatus() == HttpStatus.NOT_FOUND.value()) {
+    } catch (SpinnakerNetworkException e) {
+      throw new SystemException("getUserServiceAccounts failed", e).setRetryable(true);
+    } catch (SpinnakerHttpException e) {
+      if (e.getResponseCode() == 404) {
         return List.of();
       }
-      boolean shouldRetry =
-          response == null || HttpStatus.valueOf(response.getStatus()).is5xxServerError();
-      throw new SystemException("getUserServiceAccounts failed", re).setRetryable(shouldRetry);
+      if (HttpStatus.valueOf(e.getResponseCode()).is5xxServerError()) {
```

Comment:
Isn't it closer to the original logic to keep a shouldRetry variable and always throw a SystemException?  The original logic didn't call UpstreamBadRequest.classifyError from this method.

Reply:
Thanks.  There's a lot of duplication now though.  If we keep separate catch blocks for the specific types, the code here could become:
```
boolean shouldRetry = HttpStatus.valueOf(response.getStatus()).is5xxServerError();
throw new SystemException("getUserServiceAccounts failed", e).setRetryable(shouldRetry);
```
but I wonder if it isn't simpler to have only one catch block something like:
```
catch (SpinnakerServerException e) {
  throw new SystemException("getUserServiceAccounts failed", e).setRetryable(e.getRetryable());
}
```
This does mean some other code will need to handle null retryable, and I think is a bit of a behavior change in the http response case.  In the old logic, only 5xx is retryable.  If we do it this way, based on [this code in SpinnakerRetrofitErrorHandler](https://github.com/spinnaker/kork/blob/cf76e731d4201d0972cec6197a0f2536046e4537/kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/exceptions/SpinnakerRetrofitErrorHandler.java#L55-L59), everything except 400 and 404 is retryable.  So, perhaps the best thing is a combination...two catch blocks, one for SpinnakerHttpException and one for SpinnakerServerException.

Reply:
So we will let some other code handle null retryable now?

Reply:
RetrySupport already handles null retryable.  What other code are we talking about?

Reply:
How about this:
```
    } catch (SpinnakerHttpException e) {
      if (e.getResponseCode() == 404) {
        return List.of();
      }
      boolean shouldRetry = HttpStatus.valueOf(e.getResponseCode()).is5xxServerError();
      throw new SystemException("getUserServiceAccounts failed", e).setRetryable(shouldRetry);
    } catch (SpinnakerServerException e) {
      boolean shouldRetry = !(e instanceof SpinnakerConversionException);
      throw new SystemException("getUserServiceAccounts failed", e).setRetryable(shouldRetry);
    }
```

Reply:
Yes, that's better, but I still don't understand why we need the special case for conversion exception / why null retryable doesn't work.

Reply:
This way conversion exception need not be retried. But if it's ok to be retried though the result doesn't change, I can leave retryable null for all cases within SpinnakerServerException catch block.

Reply:
Doesn't
```
  throw new SystemException("getUserServiceAccounts failed", e).setRetryable(e.getRetryable());
```
handle this?  As in, conversion exceptions have retryable set to false.

Reply:
ahh, yes. I was not checking that part.

Reply:
Thanks for wading through that with me!